### PR TITLE
docs: Composition examples

### PR DESCRIPTION
**Pull request by johanandren**
Added scala shopping cart value entity project

A few samples replaced + added
 * new sample for forwarding using the shopping cart (slightly more realistic but still not great, keeping the samples to only be about the shopping cart though) 
 * async call + mapped response for id generation
 * composing a sequence of calls
 * counter-example to show case the lack of transaction/boundary for an action/composition

Also fixed integration test support for the testkit (dynamic port for proxy needed to come from test containers)

### CODE REVIEW

File: docs/src/modules/java/pages/actions-as-controller.adoc [hunk 1]
```diff
@@ -3,86 +3,177 @@
...
-<1> Import the Counter API definition
-<2> The protobuf option (akkaserverless.service) is specific to code-generation as provided by the Akka Serverless sbt plugin. This annotation indicates to the code-generation that an Action must be generated.
-<3> The Action mimics the Counter API
+<1> check if the added item is carrots
+<2> if it is "carrots" immediately return an error, disallowing adding the item
+<3> for allowed requests, use `components.shoppingCart.addItem()` to get a `DeferredCall`
+<4> The `DeferredCall` is used with `effects.forward()` to forward the request to the entity
+
+
+== Async Replies (FIXME better title)
```
Comment (patriknw) [inline docs/src/modules/java/pages/actions-as-controller.adoc:45]: Transform response from DeferredCall

?

File: docs/src/modules/java/pages/actions-as-controller.adoc [hunk 2]
```diff
@@ -3,86 +3,177 @@
...
 
-== Implementing the Action
+The `asyncReply` and `asyncEffect` effects allow us to accept a call, process and transform the request, pass it or a
+new request on to another component, and when that component responds also process and transform the response.
```
Comment (patriknw) [inline docs/src/modules/java/pages/actions-as-controller.adoc:48]: "pass it or a new request on to another component" makes the sentence complicated. "call another component" ?

**Review by patriknw — APPROVED** [review]
LGTM

File: docs/src/modules/java/pages/actions-as-controller.adoc (hunk 1 above)
Comment (johanandren) [reply docs/src/modules/java/pages/actions-as-controller.adoc]: Went with "Transform Request and Response to Another Component"

File: docs/src/modules/java/pages/actions-as-controller.adoc [hunk 3]
```diff
@@ -3,86 +3,176 @@
...
+
+== Transform Request and Response to Another Component
 
-== Implementing the Action
+The `asyncReply` and `asyncEffect` effects allow us to process and transform a request before calling another component and then also transform the response.
 
-The class `DoubleCounterAction` gets generated for us based on the proto file defined above.
+As an example, let us look at the problem of creating a new entity with an id generated by the service itself.
+
+In this example we implement a `Initialize` command for the controller Action which returns the message `NewCartCreated`
```
Comment (pvlugter) [inline docs/src/modules/java/pages/actions-as-controller.adoc:51]:
```suggestion
In this example we implement an `Initialize` command for the controller Action which returns the message `NewCartCreated`
```

File: docs/src/modules/java/pages/actions-as-controller.adoc [hunk 4]
```diff
@@ -3,86 +3,176 @@
...
+<1> We generate a new UUID
+<2> We use `components.shoppingCart.create(...)` to create a `DeferredCall` for `create` on the shopping cart
+<3> `execute()` on the `DeferredCall` immediately triggers a call and returns a `Future` for the response
+<4>  On a successful response, we `map` the `Empty` reply to a reply effect with the reply `NewCartCreated`
+<5> If the call leads to an error, we `recover` and return an error effect asking the client to retry
+<6> `effects.asyncEffect()` allows us to reply with a `Future[Effect[NewCartCreated]]` rather than a reply we already have created
+
+The action generates a link:https://en.wikipedia.org/wiki/Universally_unique_identifier[UUID] to use as entity id for the shopping cart. UUIDs are extremely unlikely to lead to the same id
+being generated, but to completely guarantee two calls can never be assigned the same shopping cart we
+make use of on the "boundary of consistency" provided by the entity - the entity will only process a single command at a time
```
Comment (pvlugter) [inline docs/src/modules/java/pages/actions-as-controller.adoc:87]:
```suggestion
make use of the "boundary of consistency" provided by the entity - the entity will only process a single command at a time
```

File: docs/src/modules/java/pages/actions-as-controller.adoc [hunk 5]
```diff
@@ -3,86 +3,176 @@
...
 Scala::
 +
+.src/main/java/com/example/shoppingcart/ShoppingCartActionImpl.scala
 [source,scala]
-./src/main/scala/com/example/fibonacci/Main.scala
 ----
-include::example$scala-valueentity-counter/src/main/scala/com/example/Main.scala[tag=registration]
+include::example$scala-valueentity-shopping-cart/src/main/scala/com/example/shoppingcart/ShoppingCartActionImpl.scala[tag=createPrePopulated]
+----
+<1> For comprehensions (or directly using `mapAsync`) allow us to compose the individual async steps returning `Future`
```
Comment (pvlugter) [inline docs/src/modules/java/pages/actions-as-controller.adoc:140]: `mapAsync` should be `flatMap`?

```suggestion
<1> For comprehensions (or directly using `flatMap`) allow us to compose the individual async steps returning `Future`
```

File: docs/src/modules/java/pages/actions-as-controller.adoc [hunk 6]
```diff
@@ -3,86 +3,176 @@
...
-./src/main/java/com/example/Main.java
 ----
-include::example$java-valueentity-counter/src/main/java/com/example/Main.java[tag=registration]
+include::example$java-valueentity-shopping-cart/src/main/java/com/example/shoppingcart/ShoppingCartActionImpl.java[tag=createPrePopulated]
 ----
+<1> `CompletionStage#thenCompose` allow us to perform an additional async operation, returning a `CompletionStage` once the current one completes successfully
+<2> Create a request to add an initial item to the cart
+<3> Executing the `addItem` call returns a `CompletionStage<Empty>` once it succeeds
+<4> `handle` allows us to transform the successful completion of `addItem` with `Empty` to the response type of this method - `NewCartCreated`
+<5> `effects().asyncReply()` lets us reply once the `CompletionStage<NewCartCreated>` completes.
```
Comment (pvlugter) [inline docs/src/modules/java/pages/actions-as-controller.adoc:131]: tiny thing, but be consistent with no period at the end of the callout comments?

```suggestion
<5> `effects().asyncReply()` lets us reply once the `CompletionStage<NewCartCreated>` completes
```

File: docs/src/modules/java/pages/actions-as-controller.adoc [hunk 7]
```diff
@@ -3,86 +3,176 @@
...
 +
+.src/main/java/com/example/shoppingcart/ShoppingCartActionImpl.scala
 [source,scala]
-./src/main/scala/com/example/fibonacci/Main.scala
 ----
-include::example$scala-valueentity-counter/src/main/scala/com/example/Main.scala[tag=registration]
+include::example$scala-valueentity-shopping-cart/src/main/scala/com/example/shoppingcart/ShoppingCartActionImpl.scala[tag=createPrePopulated]
+----
+<1> For comprehensions (or directly using `mapAsync`) allow us to compose the individual async steps returning `Future`
+<2> Once both steps has completed, create a `NewCartCreated` leading to a `Future[NewCartCreated]` coming out of the for-comprehension
```
Comment (pvlugter) [inline docs/src/modules/java/pages/actions-as-controller.adoc:141]:
```suggestion
<2> Once both steps have completed, create a `NewCartCreated` leading to a `Future[NewCartCreated]` coming out of the for-comprehension
```

File: docs/src/modules/java/pages/actions-as-controller.adoc [hunk 8]
```diff
@@ -3,86 +3,176 @@
...
+<3> `effect.asyncReply` lets us reply once the `Future[NewCartCreated]` completes.
```
Comment (pvlugter) [inline docs/src/modules/java/pages/actions-as-controller.adoc:142]:
```suggestion
<3> `effect.asyncReply` lets us reply once the `Future[NewCartCreated]` completes
```

File: docs/src/modules/java/pages/actions-as-controller.adoc [hunk 9]
```diff
@@ -3,86 +3,176 @@
...
+[.tabset]
+Java::
++
+.src/main/java/com/example/shoppingcart/ShoppingCartActionImpl.java
+[source,java]
 ----
+include::example$java-valueentity-shopping-cart/src/main/java/com/example/shoppingcart/ShoppingCartActionImpl.java[tag=unsafeValidation]
+----
+<1> Between this call returning
+<2> And this next call to the same entity, the entity could accept other commands change the total count of items in the cart
```
Comment (pvlugter) [inline docs/src/modules/java/pages/actions-as-controller.adoc:163]:
```suggestion
<2> And this next call to the same entity, the entity could accept other commands that change the total count of items in the cart
```

File: samples/scala-valueentity-shopping-cart/README.md [hunk 10]
```diff
@@ -0,0 +1,88 @@
+# Shopping Cart example (using a Replicated Entity)
```
Comment (pvlugter) [inline samples/scala-valueentity-shopping-cart/README.md:1]:
```suggestion
# Shopping Cart example (using a Value Entity)
```

File: samples/scala-valueentity-shopping-cart/README.md [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+This example project implements an API for a shopping cart using an Akka Serverless Replicated Entity.
```
Comment (pvlugter) [inline samples/scala-valueentity-shopping-cart/README.md:3]:
```suggestion
This example project implements an API for a shopping cart using an Akka Serverless Value Entity.
```

**Review by pvlugter — APPROVED** [review]
LGTM. Useful examples.

**Pull request merged**